java/server/src/org/openqa/selenium/grid/sessionqueue/NewSessionQueuer.java

```diff
@@ -0,0 +1,132 @@
+import org.openqa.selenium.internal.Require;
+import org.openqa.selenium.remote.NewSessionPayload;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+import org.openqa.selenium.remote.http.Routable;
+import org.openqa.selenium.remote.http.Route;
+import org.openqa.selenium.remote.tracing.AttributeKey;
+import org.openqa.selenium.remote.tracing.EventAttribute;
+import org.openqa.selenium.remote.tracing.EventAttributeValue;
+import org.openqa.selenium.remote.tracing.Span;
+import org.openqa.selenium.remote.tracing.Tracer;
+import org.openqa.selenium.status.HasReadyState;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.logging.Logger;
+
+public abstract class NewSessionQueuer implements HasReadyState, Routable {
+
+  private static final Logger LOG = Logger.getLogger(NewSessionQueuer.class.getName());
+  private final Route routes;
+  protected final Tracer tracer;
+
+  protected NewSessionQueuer(Tracer tracer) {
+    this.tracer = Require.nonNull("Tracer", tracer);
+
+    routes = combine(
+        post("/session")
+            .to(() -> req -> addToQueue(req)),
+        post("/se/grid/newsessionqueuer/session")
+            .to(() -> new AddToSessionQueue(tracer, this)),
+        post("/se/grid/newsessionqueuer/session/retry/{requestId}")
+            .to(params -> new AddBackToSessionQueue(tracer, this,
+                                                    UUID.fromString(params.get("requestId")))),
+        Route.get("/se/grid/newsessionqueuer/session")
+            .to(() -> new RemoveFromSessionQueue(tracer, this)));
+  }
+
+  public void validateSessionRequest(HttpRequest request) {
+    try (Span span = tracer.getCurrentContext().createSpan("newsession_queuer.validate")) {
+      Map<String, EventAttributeValue> attributeMap = new HashMap<>();
+      try (
+          Reader reader = reader(request);
+          NewSessionPayload payload = NewSessionPayload.create(reader)) {
+        Objects.requireNonNull(payload, "Requests to process must be set.");
+        attributeMap.put("request.payload", EventAttribute.setValue(payload.toString()));
+
+        Iterator<Capabilities> iterator = payload.stream().iterator();
+        if (!iterator.hasNext()) {
+          SessionNotCreatedException
+              exception =
+              new SessionNotCreatedException("No capabilities found");
+          EXCEPTION.accept(attributeMap, exception);
+          attributeMap.put(AttributeKey.EXCEPTION_MESSAGE.getKey(),
+                           EventAttribute.setValue(exception.getMessage()));
+          span.addEvent(AttributeKey.EXCEPTION_EVENT.getKey(), attributeMap);
+          throw exception;
+        } else {
+          Capabilities caps = iterator.next();
+          if (caps.getBrowserName().isEmpty()) {
```

Comment:
This assumes that there won't be a slot stereotype that checks for something different, and we can't be sure of that. The best we can do for validation is a fast pass through the stereotypes the `Distributor` currently knows about, and if none of those could match the `Capabilities` then we _may_ have an error, but only if all the `Capabilities` in the new session request are unmatchable.

Reply:
Alright. So for validation, we want to use the immutable model of hosts and see if any slots match any of the requested capabilities, if not we want to send back an error response?